src/Raven.Client/Documents/AI/Chat.cs

```diff
@@ -0,0 +1,132 @@
+using System.Threading.Tasks;
+using Raven.Client.Documents.Operations;
+using Raven.Client.Documents.Operations.AI.Agents;
+using Raven.Client.Util;
+
+namespace Raven.Client.Documents.AI;
+internal class Chat<T> : IChatOperations<T> where T : new()
+{
+    private readonly AiOperations _aiOperations;
+    private readonly string _agent;
+    private readonly Dictionary<string, object> _scope;
+
+    private string _chatId;
+    private AiUsage _totalUsage;
+    private List<ToolRequest> _toolsRequests;
+    private List<ToolResponse> _toolsResponses = [];
+    private string _userPrompt;
+
+    private bool _firstRun = true;
+    public Chat(AiOperations aiOperations, string agent, Dictionary<string, object> scope)
+    {
+        _aiOperations = aiOperations;
+        _agent = agent;
+        _scope = scope;
+    }
+    public Chat(AiOperations aiOperations, string chatId)
+    {
+        _aiOperations = aiOperations;
+        _chatId = chatId;
+    }
+
+    public IEnumerable<ToolRequest> OpenTools() => _toolsRequests ?? throw new InvalidOperationException("You have to call RunAsync first");
+
+    public void AddToolResponse(string id, string content)
+    {
+        _toolsResponses.Add(new ToolResponse
+        {
+            ToolId = id,
+            Content = content
+        });
+    }
+
+    public bool Run() => AsyncHelpers.RunSync(() => RunAsync(CancellationToken.None));
+
+    public void SetPrompt(string userPrompt)
+    {
+        _userPrompt = userPrompt;
+    }
+
+    private T _answer;
+    public T Answer => _answer ?? throw new InvalidOperationException("You have to call RunAsync first");
+    public string Id => _chatId ?? throw new InvalidOperationException("This is a new chat, the ID wasn't set yet, you have to call RunAsync");
+    public AiUsage TotalUsage => _totalUsage ?? throw new InvalidOperationException("You have to call RunAsync first");
+
+    public async Task<IEnumerable<ChatMessage>> ReadMessagesAsync(CancellationToken token)
+    {
+        var id = _chatId ?? throw new InvalidOperationException("This is a new chat, the ID wasn't set yet, you have to call RunAsync.");
+        using var session = _aiOperations._store.OpenAsyncSession();
+        var d = await session.LoadAsync<ChatDocument>(id, token).ConfigureAwait(false);
+        return d.Messages;
+    }
+
+    public async Task<bool> RunAsync(CancellationToken token)
+    {
+        // clear to avoid reusing old chat answer
+        _answer = default;
+
+        IMaintenanceOperation<ChatResult<T>> op;
+        if (string.IsNullOrWhiteSpace(_chatId))
+        {
+            op = new RunChatOperation<T>(_agent, _userPrompt, _scope);
+        }
+        else
+        {
+            // we allow to run the chat only if it is the first run with no user prompt or tool requests
+            // this way we can fetch the initial chat state
+            if (_firstRun == false && string.IsNullOrEmpty(_userPrompt) && _toolsResponses.Count == 0)
```

Comment:
Why do you check `_firstRun` here? Why does that matter if we run it once or not?
If you don't have a tool call to run, and no prompt - it should fail, no?